### PR TITLE
Linear model creation and interpretation

### PR DESCRIPTION
**Pull request by MattsonCam**
Here I create multiple linear models that explore multiple relationships between covariates, genotypes, and morphology features. There are other analysis I'm still considering, however, any suggestions are welcome!

### CODE REVIEW

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (gwaybio) [inline 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py:9]: Remember to make sure all code blocks are executed sequentially. (It looks like this notebook was not executed)

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (gwaybio) [inline 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py:12]: consider importing the full pathlib library

```python
import pathlib
```

It's a pretty small library and functions that execute `pathlib.Path()` are more readable than `Path()` IMO

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (gwaybio) [inline 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py:42]: I'm wondering why this is necessary? I haven't seen this logic before. Don't you execute the notebooks in the top level directory of the repo? Also, if you're using relative paths from the `.ipynb` file, shouldn't everything work fine without this logic?

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (gwaybio) [inline 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py:64]: ohh, I see - you're dynamically creating the root dir so that it works on any system. Hmm, interesting. I kind of like this approach better than the `../../` etc. approach.

In this case, consider adding a comment to the code block above stating this purpose (if my reading is correct).

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (gwaybio) [reply 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py:42]: see comment related comment below (I think I understand why you're doing this, and I provide a description below).

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (gwaybio) [inline 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py]: is this dropping rows that include missing values? Consider adding a comment describing this process

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (gwaybio) [inline 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py]: You might also consider [pandas chaining](https://towardsdatascience.com/using-pandas-method-chaining-to-improve-code-readability-d8517c5626ac) here. I love chaining in pandas!

```suggestion
cp_df = pd.concat([plate1df, plate2df], axis="rows")
cp_df = (
    cp_df
    .loc[:, (cp_df != 0).any(axis=0)]
    .reset_index(drop=True)
)
```

Comment (gwaybio) [reply 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py]: oh wait, it's probably good to not `commit` this suggestion since it's in the nbconverted file.

If you decide to make this change, it belongs in the `.ipynb`

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (gwaybio) [inline 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py:109]: i recommend checking out [`pycytominer.cyto_utils.infer_cp_features()`](https://github.com/cytomining/pycytominer/blob/master/pycytominer/cyto_utils/features.py#L78)

this may save you from reinventing wheels.

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (gwaybio) [inline 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py:131]: This is beautiful! 😍

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (gwaybio) [inline 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py]: i recommend not using a symbol here. Perhaps `neg_log_p` instead?

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (gwaybio) [inline 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py]: I generally find it very useful to define this output file at the beginning of the notebook (where you also define input paths)

This way a reader will instantly know what output (and how many outputs) to expect

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py
Comment (gwaybio) [inline 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py:74]: I recommend defining input and output paths prior to any data wrangling. It makes these easier to read!

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py
Comment (gwaybio) [inline 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py]: The significance theshold needs to be adjusted for multiple tests

so something like:

```python
# Assuming the rows of concatenated_df contain each feature
num_tests = concatenated_df.shape[0]
multiple_test_correction_alpha = 0.05 / num_tests
significance_threshold = -np.log10(multiple_test_correction_alpha)
```

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py
Comment (gwaybio) [inline 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py]: consider using color-friendly color palettes (see here for a resource: https://colorbrewer2.org/#type=sequential&scheme=BuGn&n=3)

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py
Comment (gwaybio) [inline 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py]: consider setting i/o paths at the beginning of the notebook to increase readability

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py
Comment (gwaybio) [inline 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py]: It doesn't look like you're using all of these imports. 

Consider using the `pycln` pre-commit hook to help with auto-cleaning these: https://pre-commit.com/hooks.html

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py
Comment (gwaybio) [inline 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py]: same comment about `pycln` pre-commit here

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py
Comment (gwaybio) [inline 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py:68]: same pattern about setting i/o before data science

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py
Comment (gwaybio) [inline 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py]:
```suggestion
# Create a list of all possible compartment pairs
```

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py
Comment (gwaybio) [inline 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py:74]: It might be useful here (or maybe even in the first cell of this notebook) to describe your procedure for splitting compartments in full summary detail. It will also help to provide rationale in this initial statement for why splitting is necessary.

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py
Comment (gwaybio) [inline 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py]: same comments about i/o pathing.

Also, it may be worthwhile to only output one `.tsv` file and include a boolean column (also probably the raw p values) of if the feature is significant or not.

Please also update the alpha (0.05) to accommodate multiple testing correction (see previous comment)

**Review by gwaybio — APPROVED** [review]
this LGTM! Well done.

I had several, all relatively minor, comments and suggestions. They boil down to:

1. Confirming input/output paths early in notebooks
2. Adding documentation in key places to help a reader understand.

There are also other comments regarding potential code upgrades and making sure alpha is adjusted for multiple testing.

**Comment by gwaybio** [comment]
I should also explicitly state that once a PR is approved, it can be merged, and there is no need for additional review. That being said, we expect that each scientist/engineer holds themselves to high standards and will work towards improving code and documentation. The barrier to merging a PR is primarily on the core contributor! They hold the decision power; the reviewer also has some power, but ultimately trusts the core contributor to make decisions that benefit the project readability and accuracy.

Happy to continue the discussion on any particular comment, but ultimately the decisions are up to you.

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (MattsonCam) [reply 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py:9]: Good to know, changed!

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (MattsonCam) [reply 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py:12]: I'll keep this in mind

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (MattsonCam) [reply 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py:64]: Good idea, added a description!

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (MattsonCam) [reply 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py]: This drops columns that contain all zeros, also added a comment.

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (MattsonCam) [reply 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py]: Right now I use a specially formatted python file as a jupyter notebook file, so the jupyter notebook is just for reviewers right now.

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (MattsonCam) [reply 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py:109]: I like the idea of reusing code, the one caveat with this function is that to remove the metadata you have to specify the compartments, if they don't follow the default argument, to remove metadata. The class I was using doesn't incorporate a lot of functionality, so I was also considering making it a snippet in the future.

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (MattsonCam) [reply 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py]: I thought about that, I think that's a good call. Changed!

File: 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py
Comment (MattsonCam) [reply 1.train_models/linear_reg_plate_1_2_norm_data/nbconverted/train_model.py]: Good point, changed!

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py
Comment (MattsonCam) [reply 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py:74]: Sounds good, changed!

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py
Comment (MattsonCam) [reply 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py]: I decided to use Benjamini-Hochberg correction, since it has a greater power and still controls for FDR. I think we want a less conservative approach here, since we are interested in which features could be important, rather than which features are not important. What do you think?

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py
Comment (MattsonCam) [reply 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py]: Changed!

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py
Comment (MattsonCam) [reply 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py]: Changed!

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py
Comment (MattsonCam) [reply 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py]: That's strange, I think I've figured out a work-around for the time being, and it should be updated now.

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py
Comment (MattsonCam) [reply 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py:68]: Changed!

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py
Comment (MattsonCam) [reply 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py]: Changed!

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py
Comment (MattsonCam) [reply 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py:74]: Good idea, I just added a statement.

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py
Comment (MattsonCam) [reply 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py]: I ended up using a temporary workaround for this, I'm going to try vulture.

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py
Comment (MattsonCam) [reply 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/seperate_compartments.py]: Thanks for catching that, changed! I also commented above.

**Comment by MattsonCam** [comment]
Thanks for reviewing @gwaybio, very helpful! I just had one unresolved conversation I wanted to get your opinion on

File: 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py
Comment (gwaybio) [reply 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py]: interesting! I did not see where you applied BH. Could you make that more explicit with a comment? Also, is it fair to keep the dotted line in the volcano to indicate significance then?

Comment (MattsonCam) [reply 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py]: Right! The significance threshold would change, but it doesn't change very much, about a 0.0004 difference with an FDR level of 0.05. I'll include this adjusted significance threshold, for completeness. 

BH is applied in the training step after the original p_values are calculated. I'll also add more documentation

Comment (MattsonCam) [reply 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py]: Just made the updates @gwaybio, I think an FDR level of 0.05 will suffice, I've heard it is a common value to use...

Comment (gwaybio) [reply 3.interpret_models/linear_reg_plate_1_2_norm_data/feature_importances/nbconverted/make_volcano.py]: Sounds good! Feel free to resolve the conversation and merge when you're happy

**Pull request merged**
